File: lsp/src/pyproject-hover.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import * as child_process from 'child_process';
import * as https from 'https';
import * as vscode from 'vscode';
import {PythonEnvironment} from './python-environment';

type Dependency = {
  name: string;
  configuredVersion?: string;
  range: vscode.Range;
};

type PypiMetadata = {
  summary?: string;
  version?: string;
  publishedAt?: Date;
  homepage?: string;
};

const pypiMetadataCache = new Map<string, Promise<PypiMetadata | undefined>>();
const installedVersionCache = new Map<string, Promise<string | undefined>>();

export function registerPyprojectHoverProvider(
  context: vscode.ExtensionContext,
  pythonEnv: PythonEnvironment,
) {
  context.subscriptions.push(
    vscode.languages.registerHoverProvider(
      [
        {scheme: 'file', language: 'toml', pattern: '**/pyproject.toml'},
        {scheme: 'untitled', language: 'toml'},
      ],
      {
        async provideHover(document, position, token) {
          if (
            document.uri.scheme === 'file' &&
            document.uri.fsPath.split(/[\\/]/).pop() !== 'pyproject.toml'
          ) {
            return undefined;
          }
          const dependency = findPyprojectDependency(document, position);
          if (!dependency) {
            return undefined;
          }
          const [metadata, installedVersion] = await Promise.all([
            fetchPypiMetadata(dependency.name),
            resolveInstalledVersion(pythonEnv, document.uri, dependency.name),
          ]);
          if (token.isCancellationRequested) {
            return undefined;
          }
          const markdown = formatHover(dependency, metadata, installedVersion);
          return markdown ? new vscode.Hover(markdown, dependency.range) : undefined;
        },
      },
    ),
  );
}

export function findPyprojectDependency(
  document: vscode.TextDocument,
  position: vscode.Position,
): Dependency | undefined {
  const line = document.lineAt(position.line).text;
  const section = findSection(document, position.line);
  const quoted = dependencyFromQuotedString(line, position);
  if (quoted) {
    const arrayKey = findEnclosingArrayKey(document, position.line);
    if (
      (section === 'project' && arrayKey === 'dependencies') ||
      section === 'project.optional-dependencies' ||
      section === 'dependency-groups'
    ) {
      return quoted;
    }
  }
  if (
    section === 'tool.poetry.dependencies' ||
    (section?.startsWith('tool.poetry.group.') &&
      section.endsWith('.dependencies'))
  ) {
    return dependencyFromPoetryEntry(line, position);
  }
  return undefined;
}

function findSection(document: vscode.TextDocument, line: number): string | undefined {
  for (let i = line; i >= 0; i--) {
    const match = document.lineAt(i).text.match(/^\s*\[([^\]]+)\]\s*(?:#.*)?$/);
    if (match) {
      return match[1].trim();
    }
  }
  return undefined;
}

function findEnclosingArrayKey(
  document: vscode.TextDocument,
  line: number,
): string | undefined {
  for (let i = line; i >= 0; i--) {
    const text = document.lineAt(i).text;
    if (/^\s*\[/.test(text)) {
      return undefined;
    }
    const match = text.match(/^\s*([A-Za-z0-9_.-]+)\s*=\s*\[/);
    if (match) {
      return match[1];
    }
    if (i < line && text.includes(']')) {
      return undefined;
    }
  }
  return undefined;
}

function dependencyFromQuotedString(
  line: string,
  position: vscode.Position,
): Dependency | undefined {
  for (const span of quotedSpans(line)) {
    if (position.character < span.start || position.character > span.end) {
      continue;
    }
    const text = line.slice(span.start + 1, span.end);
    const parsed = parseDependencySpec(text);
    if (!parsed) {
      return undefined;
    }
    return {
      ...parsed,
      range: new vscode.Range(
        position.line,
        span.start + 1,
        position.line,
        span.start + 1 + parsed.name.length,
      ),
    };
  }
  return undefined;
}

function quotedSpans(line: string): {start: number; end: number}[] {
  const spans = [];
  let quote: string | undefined;
  let start = -1;
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (quote) {
      if (char === quote && line[i - 1] !== '\\') {
        spans.push({start, end: i});
        quote = undefined;
      }
    } else if (char === '"' || char === "'") {
      quote = char;
      start = i;
    }
  }
  return spans;
}

function dependencyFromPoetryEntry(
  line: string,
  position: vscode.Position,
): Dependency | undefined {
  const match = line.match(/^\s*("[^"]+"|'[^']+'|[A-Za-z0-9_.-]+)\s*=\s*(.+)$/);
  if (!match) {
    return undefined;
  }
  const rawName = match[1];
  const keyStart = line.indexOf(rawName);
  const keyEnd = keyStart + rawName.length;
  const name = rawName.replace(/^["']|["']$/g, '');
  if (name === 'python') {
    return undefined;
  }
  const value = match[2].trim();
  const range = new vscode.Range(position.line, keyStart, position.line, keyEnd);
  if (position.character < keyStart || position.character > line.length) {
    return undefined;
  }
  return {
    name,
    configuredVersion: poetryVersion(value),
    range,
  };
}

function parseDependencySpec(
  spec: string,
): {name: string; configuredVersion?: string} | undefined {
  const match = spec
    .trim()
    .match(/^([A-Za-z0-9][A-Za-z0-9._-]*)(?:\s*\[[^\]]+\])?\s*(.*)$/);
  if (!match || match[1] === 'python') {
    return undefined;
  }
  return {
    name: match[1],
    configuredVersion: exactVersion(match[2]),
  };
}

function exactVersion(spec: string): string | undefined {
  return spec.match(/^\s*==\s*([^,;\s]+)/)?.[1];
}

function poetryVersion(value: string): string | undefined {
  const quoted = value.match(/^["']([^"']+)["']/)?.[1];
  if (quoted) {
    return quoted;
  }
  return value.match(/version\s*=\s*["']([^"']+)["']/)?.[1];
}

async function fetchPypiMetadata(name: string): Promise<PypiMetadata | undefined> {
  const normalized = normalizePackageName(name);
  let cached = pypiMetadataCache.get(normalized);
  if (!cached) {
    cached = fetchPypiMetadataUncached(name);
    pypiMetadataCache.set(normalized, cached);
  }
  return cached;
}

function fetchPypiMetadataUncached(
  name: string,
): Promise<PypiMetadata | undefined> {
  return new Promise(resolve => {
    const request = https.get(
      {
        hostname: 'pypi.org',
        path: `/pypi/${encodeURIComponent(name)}/json`,
        headers: {'User-Agent': 'pyrefly-vscode'},
        timeout: 3000,
      },
      response => {
        if (response.statusCode !== 200) {
          response.resume();
          resolve(undefined);
          return;
        }
        let body = '';
        response.setEncoding('utf8');
        response.on('data', chunk => {
          body += chunk;
        });
        response.on('end', () => {
          try {
            const json = JSON.parse(body);
            const version = json.info?.version;
            const releases = version ? json.releases?.[version] : undefined;
            const publishedAt = Array.isArray(releases)
              ? releases
                  .map(release => release.upload_time_iso_8601)
                  .filter(Boolean)
                  .sort()
                  .pop()
              : undefined;
            resolve({
              summary: json.info?.summary,
              version,
              publishedAt: publishedAt ? new Date(publishedAt) : undefined,
              homepage:
                json.info?.project_urls?.Homepage ??
                json.info?.project_urls?.Source ??
                json.info?.home_page ??
                json.info?.package_url,
            });
          } catch {
            resolve(undefined);
          }
        });
      },
    );
    request.on('timeout', () => {
      request.destroy();
      resolve(undefined);
    });
    request.on('error', () => resolve(undefined));
  });
}

async function resolveInstalledVersion(
  pythonEnv: PythonEnvironment,
  uri: vscode.Uri,
  name: string,
): Promise<string | undefined> {
  const interpreter = await pythonEnv.getInterpreterPath(uri);
  if (!interpreter) {
    return undefined;
  }
  const key = `${interpreter}\0${normalizePackageName(name)}`;
  let cached = installedVersionCache.get(key);
  if (!cached) {
    cached = queryInstalledVersion(interpreter, name);
    installedVersionCache.set(key, cached);
  }
  return cached;
}

function queryInstalledVersion(
  interpreter: string,
  name: string,
): Promise<string | undefined> {
  const script = [
    'import importlib.metadata',
    'import sys',
    'try:',
    '    print(importlib.metadata.version(sys.argv[1]))',
    'except importlib.metadata.PackageNotFoundError:',
    '    pass',
  ].join('\n');
  return new Promise(resolve => {
    child_process.execFile(
      interpreter,
      ['-c', script, name],
      {timeout: 2000},
      (error, stdout) => {
        if (error) {
          resolve(undefined);
          return;
        }
        const version = stdout.trim();
        resolve(version === '' ? undefined : version);
      },
    );
  });
}

function formatHover(
  dependency: Dependency,
  metadata?: PypiMetadata,
  installedVersion?: string,
): vscode.MarkdownString | undefined {
  if (!metadata && !installedVersion && !dependency.configuredVersion) {
    return undefined;
  }
  const markdown = new vscode.MarkdownString(undefined, true);
  const summary = metadata?.summary?.trim();
  if (summary) {
    markdown.appendText(summary);
  }
  if (metadata?.version) {
    appendParagraph(
      markdown,
      `Latest version: ${metadata.version}${publishedSuffix(metadata.publishedAt)}`,
    );
  }
  if (installedVersion) {
    appendParagraph(markdown, `Installed version: ${installedVersion}`);
  } else if (dependency.configuredVersion) {
    appendParagraph(markdown, `Configured version: ${dependency.configuredVersion}`);
  }
  if (metadata?.homepage) {
    appendParagraph(markdown, metadata.homepage);
  }
  return markdown;
}

function appendParagraph(markdown: vscode.MarkdownString, text: string) {
  if (markdown.value.length > 0) {
    markdown.appendMarkdown('\n\n');
  }
  markdown.appendText(text);
}

function publishedSuffix(date?: Date): string {
  if (!date || Number.isNaN(date.getTime())) {
    return '';
  }
  return ` published ${relativeTime(date)}`;
}

function relativeTime(date: Date): string {
  const seconds = Math.max(1, Math.floor((Date.now() - date.getTime()) / 1000));
  const units: [string, number][] = [
    ['year', 365 * 24 * 60 * 60],
    ['month', 30 * 24 * 60 * 60],
    ['week', 7 * 24 * 60 * 60],
    ['day', 24 * 60 * 60],
    ['hour', 60 * 60],
    ['minute', 60],
  ];
  for (const [unit, size] of units) {
    const amount = Math.floor(seconds / size);
    if (amount >= 1) {
      return `${amount} ${unit}${amount === 1 ? '' : 's'} ago`;
    }
  }
  return `${seconds} seconds ago`;
}

function normalizePackageName(name: string): string {
  return name.toLowerCase().replace(/[-_.]+/g, '-');
}

File: pyrefly/lib/lsp/non_wasm/pyproject_hover.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::io::Read;
use std::path::Path;
use std::process::Command;
use std::process::Stdio;
use std::sync::Arc;
use std::thread;
use std::time::Duration;
use std::time::Instant;
use std::time::SystemTime;
use std::time::UNIX_EPOCH;

use lsp_types::Hover;
use lsp_types::HoverContents;
use lsp_types::MarkupContent;
use lsp_types::MarkupKind;
use lsp_types::Position;
use pyrefly_config::pyproject::PyProjectDependency;
use pyrefly_config::pyproject::dependency_at_offset;
use pyrefly_util::lined_buffer::LinedBuffer;
use ruff_text_size::TextRange;
use ruff_text_size::TextSize;
use serde::Deserialize;

use crate::lsp::wasm::hover::HoverResult;

const PACKAGE_QUERY_TIMEOUT: Duration = Duration::from_secs(3);
// The selected interpreter provides both environment metadata and HTTPS without a new Rust client.
const PACKAGE_QUERY: &str = r#"
import importlib.metadata
import json
import sys
import urllib.parse
import urllib.request
from datetime import datetime

name = sys.argv[1]
result = {}
try:
    result["installed_version"] = importlib.metadata.version(name)
except importlib.metadata.PackageNotFoundError:
    pass
try:
    request = urllib.request.Request(
        "https://pypi.org/pypi/" + urllib.parse.quote(name, safe="") + "/json",
        headers={"User-Agent": "pyrefly-lsp"},
    )
    with urllib.request.urlopen(request, timeout=2) as response:
        payload = json.load(response)
    info = payload.get("info") or {}
    result["summary"] = info.get("summary")
    result["version"] = info.get("version")
    urls = info.get("project_urls") or {}
    result["homepage"] = (
        urls.get("Homepage")
        or urls.get("Source")
        or info.get("home_page")
        or info.get("package_url")
    )
    releases = (payload.get("releases") or {}).get(info.get("version"), [])
    uploads = [entry.get("upload_time_iso_8601") for entry in releases]
    uploads = [upload for upload in uploads if upload]
    if uploads:
        result["published_at"] = int(
            datetime.fromisoformat(max(uploads).replace("Z", "+00:00")).timestamp()
        )
except Exception:
    pass
print(json.dumps(result))
"#;

#[derive(Clone, Debug, Default, Deserialize)]
pub(super) struct PackageMetadata {
    summary: Option<String>,
    version: Option<String>,
    published_at: Option<u64>,
    homepage: Option<String>,
    installed_version: Option<String>,
}

/// Query local and package-index metadata with the selected Python environment.
pub(super) fn package_metadata(interpreter: &Path, name: &str) -> PackageMetadata {
    let mut child = match Command::new(interpreter)
        .args(["-c", PACKAGE_QUERY, name])
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
    {
        Ok(child) => child,
        Err(_) => return PackageMetadata::default(),
    };
    let start = Instant::now();
    loop {
        match child.try_wait() {
            Ok(Some(_)) => break,
            Ok(None) if start.elapsed() < PACKAGE_QUERY_TIMEOUT => {
                thread::sleep(Duration::from_millis(10));
            }
            Ok(None) | Err(_) => {
                let _ = child.kill();
                let _ = child.wait();
                return PackageMetadata::default();
            }
        }
    }
    let mut output = String::new();
    if child
        .stdout
        .take()
        .is_none_or(|mut stdout| stdout.read_to_string(&mut output).is_err())
    {
        return PackageMetadata::default();
    }
    serde_json::from_str(&output).unwrap_or_default()
}

/// Find the dependency under an LSP position.
pub(super) fn dependency(source: &str, position: Position) -> Option<PyProjectDependency> {
    let buffer = LinedBuffer::new(Arc::new(source.to_owned()));
    let offset = buffer.from_lsp_position(position, None).to_usize();
    dependency_at_offset(source, offset)
}

/// Build a standard LSP hover for a dependency and its available metadata.
pub(super) fn hover(
    source: &str,
    dependency: &PyProjectDependency,
    metadata: &PackageMetadata,
) -> Option<HoverResult> {
    let buffer = LinedBuffer::new(Arc::new(source.to_owned()));
    let start = TextSize::new(u32::try_from(dependency.range.start).ok()?);
    let end = TextSize::new(u32::try_from(dependency.range.end).ok()?);
    let mut paragraphs = Vec::new();
    if let Some(summary) = metadata.summary.as_deref().map(str::trim)
        && !summary.is_empty()
    {
        paragraphs.push(escape_markdown(summary));
    }
    if let Some(version) = &metadata.version {
        let published = metadata
            .published_at
            .and_then(relative_time)
            .map(|age| format!(" published {age}"))
            .unwrap_or_default();
        paragraphs.push(format!(
            "Latest version: {}{published}",
            escape_markdown(version)
        ));
    }
    if let Some(version) = &metadata.installed_version {
        paragraphs.push(format!("Installed version: {}", escape_markdown(version)));
    } else if let Some(version) = &dependency.configured_version {
        paragraphs.push(format!("Configured version: {}", escape_markdown(version)));
    }
    if let Some(homepage) = metadata.homepage.as_deref().filter(|url| {
        (url.starts_with("https://") || url.starts_with("http://"))
            && !url
                .chars()
                .any(|character| character.is_whitespace() || matches!(character, '<' | '>'))
    }) {
        paragraphs.push(format!("<{homepage}>"));
    }
    if paragraphs.is_empty() {
        return None;
    }
    Some(HoverResult {
        hover: Hover {
            contents: HoverContents::Markup(MarkupContent {
                kind: MarkupKind::Markdown,
                value: paragraphs.join("\n\n"),
            }),
            range: Some(buffer.to_lsp_range(TextRange::new(start, end), None)),
        },
        can_increase_verbosity: false,
    })
}

fn relative_time(timestamp: u64) -> Option<String> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).ok()?.as_secs();
    let seconds = now.saturating_sub(timestamp).max(1);
    for (unit, size) in [
        ("year", 365 * 24 * 60 * 60),
        ("month", 30 * 24 * 60 * 60),
        ("week", 7 * 24 * 60 * 60),
        ("day", 24 * 60 * 60),
        ("hour", 60 * 60),
        ("minute", 60),
    ] {
        let amount = seconds / size;
        if amount >= 1 {
            return Some(format!(
                "{amount} {unit}{} ago",
                if amount == 1 { "" } else { "s" }
            ));
        }
    }
    Some(format!("{seconds} seconds ago"))
}

fn escape_markdown(text: &str) -> String {
    text.chars().fold(String::new(), |mut escaped, character| {
        if matches!(
            character,
            '\\' | '`' | '*' | '_' | '{' | '}' | '[' | ']' | '<' | '>' | '#' | '|' | '!'
        ) {
            escaped.push('\\');
        }
        escaped.push(character);
        escaped
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format_dependency_hover() {
        let source = "[project]\ndependencies = [\"requests==2.32.0\"]\n";
        let result = hover(
            source,
            &dependency(source, Position::new(1, 18)).unwrap(),
            &PackageMetadata {
                summary: Some("HTTP *client*".to_owned()),
                version: Some("2.32.4".to_owned()),
                homepage: Some("https://example.com".to_owned()),
                installed_version: Some("2.32.3".to_owned()),
                ..Default::default()
            },
        )
        .unwrap();
        let HoverContents::Markup(contents) = result.hover.contents else {
            panic!("expected markdown hover")
        };
        assert_eq!(
            contents.value,
            "HTTP \\*client\\*\n\nLatest version: 2.32.4\n\nInstalled version: 2.32.3\n\n<https://example.com>"
        );
        assert_eq!(
            result.hover.range,
            Some(lsp_types::Range::new(
                Position::new(1, 17),
                Position::new(1, 25),
            ))
        );
    }

    #[test]
    fn configured_version_is_fallback() {
        let source = "[project]\ndependencies = [\"requests==2.32.0\"]\n";
        let result = hover(
            source,
            &dependency(source, Position::new(1, 18)).unwrap(),
            &PackageMetadata::default(),
        )
        .unwrap();
        let HoverContents::Markup(contents) = result.hover.contents else {
            panic!("expected markdown hover")
        };
        assert_eq!(contents.value, "Configured version: 2.32.0");
    }
}
